subsys/bluetooth/mesh/proxy_srv.c

```diff
@@ -680,155 +658,250 @@ static bool sub_count_cb(struct bt_mesh_subnet *sub, void *cb_data)
+
+	return NULL;
+}
+
+static struct {
+	int32_t start;
+	struct bt_mesh_subnet *sub;
+	struct proxy_adv_req req;
+	bool ignore_previous;
+} prev_sub_adv;
+
+static int gatt_proxy_advertise(void)
 {
-	int32_t remaining = SYS_FOREVER_MS;
-	int subnet_count;
-	int err = -EBUSY;
-	bool planned = false;
+	int err;
+
+	int32_t max_adv_duration;
+	int active_sub_cnt;
+	struct bt_mesh_subnet *sub;
+	struct proxy_adv_req request;
 
 	LOG_DBG("");
 
+	/* Close proxy activity that has timed out on all subnets */
+	bt_mesh_subnet_foreach(proxy_adv_timeout_eval);
+
 	if (!bt_mesh_proxy_has_avail_conn()) {
 		LOG_DBG("Connectable advertising deferred (max connections)");
 		return -ENOMEM;
 	}
 
-	sub = beacon_sub ? beacon_sub : bt_mesh_subnet_next(beacon_sub);
-	if (!sub) {
-		LOG_WRN("No subnets to advertise on");
+	active_sub_cnt = active_proxy_sub_cnt();
+	if (!active_sub_cnt) {
+		LOG_DBG("No subnets to advertise proxy on");
 		return -ENOENT;
-	}
-
-	subnet_count = sub_count();
-	LOG_DBG("sub_count %u", subnet_count);
-	if (subnet_count > 1) {
-		int32_t max_timeout;
+	} else if (active_sub_cnt > 1) {
+		/** There are more than one subnet that requires proxy adv,
+		 *  and the adv resources must be shared.
+		 */
 
 		/* We use NODE_ID_TIMEOUT as a starting point since it may
 		 * be less than 60 seconds. Divide this period into at least
 		 * 6 slices, but make sure that a slice is at least one
 		 * second long (to avoid excessive rotation).
 		 */
-		max_timeout = NODE_ID_TIMEOUT / MAX(subnet_count, 6);
-		max_timeout = MAX(max_timeout, 1 * MSEC_PER_SEC);
-
-		if (remaining > max_timeout || remaining == SYS_FOREVER_MS) {
-			remaining = max_timeout;
-		}
-	}
-
-	for (int i = 0; i < subnet_count; i++) {
-
-		if (sub->node_id == BT_MESH_NODE_IDENTITY_RUNNING) {
-			uint32_t active = k_uptime_get_32() - sub->node_id_start;
-			bool priv_node_id = false;
-
-			if (active < NODE_ID_TIMEOUT) {
-				remaining = MIN(remaining, NODE_ID_TIMEOUT - active);
-				LOG_DBG("Node ID active for %u ms, %d ms remaining",
-					active, remaining);
-#if defined(CONFIG_BT_MESH_PRIV_BEACONS)
-				priv_node_id = sub->priv_beacon_ctx.node_id;
-#endif
-				if (priv_node_id) {
-					err = priv_node_id_adv(sub, remaining);
-				} else {
-					err = node_id_adv(sub, remaining);
+		max_adv_duration = NODE_ID_TIMEOUT / MAX(active_sub_cnt, 6);
+		max_adv_duration = MAX(max_adv_duration, MSEC_PER_SEC);
+
+		/* Check if the previous subnet finished its allocated timeslot */
+		if ((prev_sub_adv.req.duration != SYS_FOREVER_MS) &&
+		    proxy_adv_sub_request_get(prev_sub_adv.sub, &request) &&
+		    (prev_sub_adv.req.evt == request.evt)) {
+			int32_t time_delta = k_uptime_get_32() - prev_sub_adv.start;
+
+			if (time_delta < prev_sub_adv.req.duration) {
+				sub = prev_sub_adv.sub;
+				request.duration = prev_sub_adv.req.duration - time_delta;
+				if (request.duration >= MSEC_PER_SEC) {
+					goto end;
+				} else if (!prev_sub_adv.ignore_previous) {
```

Comment:
This is because adv interval is 1 second, so if timeout is less than interval, the controller ends the advertisement immediately, right? What if we add some margin to the timeout, e.g. 10ms?

Reply:
>Can't proxy_adv_sub_request_get simply return false here

`proxy_adv_sub_request_get` returns the requested time for advertising proxy beacons from a single subnet, which is not necessarily the adv time it will be given if there are multiple subnets with proxy adv work.

Reply:
10ms is minimal because the timeout value is in 10ms step:
https://github.com/zephyrproject-rtos/zephyr/blob/ed610400bcb74338d25e4b29f948709efe12f3a3/subsys/bluetooth/mesh/adv_ext.c#L547